source/common/network/utility.h

```diff
@@ -61,11 +61,17 @@ class UdpPacketProcessor {
    * the size of datagrams received, they will be dropped.
    */
   virtual uint64_t maxDatagramSize() const PURE;
+
+  /**
+   * An estimated number of reads needed to process enough packets in each READ event.
+   */
+  virtual size_t numReadsExpectedPerEventLoop() const PURE;
 };
 
 static const uint64_t DEFAULT_UDP_MAX_DATAGRAM_SIZE = 1500;
 static const uint64_t NUM_DATAGRAMS_PER_GRO_RECEIVE = 16;
 static const uint64_t NUM_DATAGRAMS_PER_MMSG_RECEIVE = 16;
+static const uint64_t MAX_NUM_READS_PER_EVENT_LOOP = 100;
```

Comment:
What's a reasonable number for this?

Reply:
I don't know.  I would suggest running load tests that compare H2 and QUIC throughput when there are competing requests at the proxy.  I would consider config parameters that specify the number of packets or MB of QUIC data to process per worker wakeup.

Another alternative to consider if having the limit grow in cases where the UDP code notices that it wasn't able to fully drain the socket.  Or attempt to drain as many packets as they are in the UDP receive buffer at the start of the UDP receive loop.

Also consider the consequences of these limits:
Say that you have a UDP receive packet limit in the kernel of 10,000 packets but you only allow draining 100 per wakeup.  As the duration of each wakeup increases, you start getting older and older packets in each subsequent wakeup until every packet you read ends up being so old that the client has given up by the time you start processing the packet.  This can be a problem specially when handling DNS; a burst of attack DNS requests could lead to real requests backing up and failing for an extended period of time as you slowly drain old requests from the queue.

Reply:
So if I understand this, right now UDP floods can starve TCP, and if we flip this, UDP floods would starve UDP.  I agree that this should scale with session load, but as QUIC is alpha I'd prefer it be affected by starvation, and have a TODO for scaling.  This would need a relnote and a config or runtime override for non-QUIC UDP.   I think basic scaling is a fixed number of packets (was it 16?) per QUIC session if you want to address it in place.

Reply:
I'm saying that with this change, TCP will starve UDP.

Reply:
er yes, agree, mistyped.  But without it can't UDP starve TCP?

Reply:
A question please, do we have read limit for TCP connections? RawBufferSocket::doRead() doesn't seem to have it, but I'm not 100% sure.

Reply:
yeah, shouldDrainReadBuffer is how it's limited -  stops when you hit the configured buffer limit

Reply:
> So if I understand this, right now UDP floods can starve TCP, and if we flip this, UDP floods would starve UDP. I agree that this should scale with session load, but as QUIC is alpha I'd prefer it be affected by starvation, and have a TODO for scaling. This would need a relnote and a config or runtime override for non-QUIC UDP. 
For non-QUIC, do we still want to drain the socket on each read event?

> I think basic scaling is a fixed number of packets (was it 16?) per QUIC session if you want to address it in place.
If we are using recvmmsg, we get a multiplier of 16 from existing implementation. That's why numReadsExpectedPerEventLoop() just return session number.

Reply:
A possible question is how many hot fds we can get back from epoll_wait per wakeup.  I expect the answer to be somewhere in the 16 to 256 range.  For fairness between TCP and UDP on the upper side of that range with a TCP read size of 32kb per connection I think we should allow read of about 6k UDP packets per wakeup.  I guess 100 attempts returning 16 packets each is only a factor of 4 away from that, so the threshold of 100 recvmmsg with each returning up to 16 packets is actually pretty good.  I thought that the limits above resulted in 1 packet per active QUIC connection up to a limit of 100 packets.

Reply:
I changed the logic to limit number of packets read per loop and put upper limit to be 6000.